### ISSUE TITLE
`irods.test.query_test.TestQuery.test_files_query_case_sensitive`'s `between` test fails for mysql

### ISSUE THREAD

**Issue by alanking** [irods/python-irodsclient#600]
This was first discovered in https://github.com/irods/python-irodsclient/issues/597. See also: https://github.com/irods/python-irodsclient/issues/597#issuecomment-2263583927 and https://github.com/irods/irods/issues/7930

We need to be able to detect the database type being used in the iRODS server and change the assertions based on that.

**Comment by alanking** [comment]
This sub-test(ish) is now being skipped as of PR #601. A TODO linked to this issue has been left next to the code which needs to be fixed:

https://github.com/irods/python-irodsclient/blob/7af55af1f2ca8618ecf22049d4d2b26df2dcf9a2/irods/test/query_test.py#L181-L184

**Comment by d-w-moore** [comment]
Re: https://github.com/irods/python-irodsclient/blob/7af55af1f2ca8618ecf22049d4d2b26df2dcf9a2/irods/test/query_test.py#L181-L184
1. Are we better able to detect the DB flavor now?
2. What is the specific issue in the BETWEEN operator when using MySQL? What makes its behavior different to other DB's?

**Comment by korydraughn** [comment]
> 1. Are we better able to detect the DB flavor now?

No. That isn't something clients should care about. The test and/or test harness needs to provide that information. Alternatively, you can try and adjust the assertions so that the test passes regardless of the database. 

> 2. What is the specific issue in the BETWEEN operator when using MySQL? What makes its behavior different to other DB's?

That is best answered by running an experiment and/or reading the docs.
- https://dev.mysql.com/doc/refman/8.4/en/comparison-operators.html#operator_between
- https://www.postgresql.org/docs/current/functions-comparison.html

Use of the BETWEEN keyword is optional. BETWEEN can likely be replaced by explicit range checks, leading to more consistent results across database technologies. However, there may be other factors at play. For that reason, reading the docs, examining the values used in the test, running experiments, and considering the database's configuration will help in figuring out the difference.

**Comment by d-w-moore** [comment]
Thanks @korydraughn .  I did do a little experimentation. MySQL seems to fuzz the case by default, where as eg PostgreSQL does not.

**Comment by korydraughn** [comment]
It's not clear to me what you mean by "fuzz". Please explain.

**Comment by d-w-moore** [comment]
That was inaccurate of me.  By "fuzz" I really just meant the MySQL BETWEEN  is doing character comparisons/sorting case-insensitively, such that object names get sorted according to: 
'a' < 'B' < 'c'
rather than, for example:
'B' < 'a' < 'c'
as you might expect from the characters' ASCII codes.

I understand from past experience there are locales in which case insensitive ordering is enforced. Our mysql DB might, in a default installation, have such a locale set for itself; therefore its BETWEEN operator might operate on top of that altered (relative to the other DBs) sort order, rendering results differently.

**Comment by d-w-moore** [comment]
> As I stated during group discussions, I may end up moving the failing parts of the mysql test to a separate test that only runs when the client unittest is run on the same node as the server and DB. (Such a test harness exists in branch https://github.com/d-w-moore/python-irodsclient/tree/502.m

**Comment by korydraughn** [comment]
Makes sense.

That branch targets #502 which is part of the 3.2.0 milestone. Sounds like this issue should be moved into that milestone as well. You agree?

**Comment by d-w-moore** [comment]
i don't think that hurts.  I  may still try to determine now if my guesses are correct on this one, since I'm most of the way there.

**Comment by d-w-moore** [comment]
> Use of the BETWEEN keyword is optional. BETWEEN can likely be replaced by explicit range checks, leading to more consistent results across database technologies. However, there may be other factors at play. For that reason, reading the docs, examining the values used in the test, running experiments, and considering the database's configuration will help in figuring out the difference.

I'm afraid that I am ignorant in the field of SQL ... what would be an example of "explicit range checks"?

**Comment by korydraughn** [comment]
> I'm afraid that I am ignorant in the field of SQL ... what would be an example of "explicit range checks"?

It means defining the range using the order-related comparison operators (e.g. `<`, `<=`, `>`, `>=`).

**Comment by d-w-moore** [comment]
Ah, ok.  I'm betting that in a specific query as well as in GenQuery, between(a,b,c) implies / is implied by a <= b <= c or the reverse ordering being true.  But again, I'm ignorant.  I'll look to the GenQuery docs.

**Comment by korydraughn** [comment]
I assume GenQuery1 and GenQuery2 pass the BETWEEN syntax through the parser as is. Meaning, GenQuery defers the interpretation of BETWEEN to the database. We could change that in GenQuery2 allowing users to get behavior that's consistent across databases. Keep in mind that doing that could result in some features of BETWEEN being lost because the database never seeing the use of BETWEEN at the SQL level.

**Comment by korydraughn** [comment]
After some discussion, we've agreed to remove the tests or parts of tests involving the BETWEEN keyword and then update docs.irods.org to include a few words about case sensitivity in relation to databases.

I've added checkboxes to the issue to reflect the work.
